draw leaf window decorations on ->frame instead of ->parent->frame

stapelberg opened the pull request:
This fixes the graphical artifacts when gaps are enabled with title bars.

Not yet done (will update the PR):

* [x] documentation update (hacking-howto, userguide)
* [x] in `layout stacked` and `layout tabbed`, the terminal contents are one line short for some reason (wrong calculation somewhere?)

related to https://github.com/i3/i3/issues/3724

stapelberg committed:
draw_util: refactor surface_initialized macro into function

This makes it possible to set a breakpoint in gdb on a line in the function and
get a backtrace of un-initialized surface access.

stapelberg committed:
draw leaf window decorations on ->frame instead of ->parent->frame

related to https://github.com/i3/i3/issues/3724
fixes https://github.com/i3/i3/issues/1966

okraits commented:
> This fixes the graphical artifacts when gaps are enabled.

When gaps and titlebars are enabled, I guess, right?

stapelberg commented:
> When gaps and titlebars are enabled, I guess, right?

Yes

Airblader commented:
This closes #1966 I believe. I also wonder how you're feeling about https://github.com/i3/i3/issues/1966#issuecomment-158684486?

stapelberg commented:
> This closes #1966 I believe

Yep, the “fixes #1966” is already in the commit message (but not on the pull request level).

> I also wonder how you're feeling about [#1966 (comment)](https://github.com/i3/i3/issues/1966#issuecomment-158684486)?

I haven’t thought it through in detail.

On one hand, one could say that the deco_rect semantics are unchanged, only the rendering has changed — the deco_rect still specifies *where* the decoration is drawn, but now the drawing happens on a different X11 window.

I think the way we interpreted “retain semantics” in the past (at least in that issue) is to translate coordinates in such a way that the x/y position still matches the absolute position on screen where decorations appear today.

I wonder how widely used the deco_rect is by IPC interface users at all. The only usages I can find on GitHub Code Search that might be affected are https://github.com/cornerman/i3-easyfocus and https://github.com/kraftwerk28/dotfiles/blob/bd081348da1dfdc126db95563b534a7669e990bd/bin/grimvshot#L35

I’ll try and see if i3-easyfocus breaks with my change, and how hard it would be to do a translation.

stapelberg commented:
Okay, all done! Fixed the tabbed/stacked mis-calculation (within `con_border_style_rect()`), kept the existing `deco_rect` compatible (verified with i3-easyfocus that it works), updated the docs (ipc, hacking-howto, userguide).

Let me know if this looks good now :)

Airblader reviewed (approved):
🎉 In a way this change is even more meaningful than merging i3-gaps.

orestisfl commented:
I'll make a proper issue when I have time but this seems to break titlebars in my non-gaps setup: They are not showing anymore for tiled windows.

stapelberg commented:
> I'll make a proper issue when I have time but this seems to break titlebars in my non-gaps setup: They are not showing anymore for tiled windows.

Hmm, I can’t reproduce this unfortunately.

I just turned off gaps, but title bars still show up for me, both for a stacked/tabbed split container, and also in a splitv/splith normal border container. 

Can you test as a separate user to see if it also happens with the default config? 

Thanks

orestisfl commented:
It seems connected to `hide_edge_borders smart`. At least, if I delete the configuration and fallback to the default there is no issue.

stapelberg commented:
> It seems connected to `hide_edge_borders smart`. At least, if I delete the configuration and fallback to the default there is no issue.

I tried it with the default config and `hide_edge_borders smart` added to it, but everything seems to work as expected (or I’m trying to reproduce it incorrectly)
